x-pack/plugins/global_search/server/services/search_service.ts

```diff
@@ -0,0 +1,119 @@
+import { getContextFactory, GlobalSearchContextFactory } from './context';
+import {
+  GlobalSearchResultProvider,
+  GlobalSearchBatchedResults,
+  GlobalSearchFindOptions,
+} from '../types';
+
+/** @public */
+export interface SearchServiceSetup {
+  registerResultProvider(provider: GlobalSearchResultProvider): void;
+}
+
+/** @public */
+export interface SearchServiceStart {
+  find(
+    term: string,
+    options: GlobalSearchFindOptions,
+    request: KibanaRequest
+  ): Observable<GlobalSearchBatchedResults>;
+}
+
+interface SetupDeps {
+  basePath: IBasePath;
+  licenseChecker: ILicenseChecker;
+  config: GlobalSearchConfigType;
+  maxProviderResults?: number;
+}
+
+const defaultMaxProviderResults = 20;
+
+/** @internal */
+export class SearchService {
+  private readonly providers: GlobalSearchResultProvider[] = [];
+  private basePath?: IBasePath;
+  private config?: GlobalSearchConfigType;
+  private contextFactory?: GlobalSearchContextFactory;
+  private licenseChecker?: ILicenseChecker;
+  private maxProviderResults = defaultMaxProviderResults;
+
+  setup({
+    basePath,
+    config,
+    licenseChecker,
+    maxProviderResults = defaultMaxProviderResults,
+  }: SetupDeps): SearchServiceSetup {
+    this.basePath = basePath;
+    this.config = config;
+    this.licenseChecker = licenseChecker;
+    this.maxProviderResults = maxProviderResults;
+
+    return {
+      registerResultProvider: (provider) => {
+        if (this.providers.map((p) => p.id).includes(provider.id)) {
+          throw new Error(`trying to register duplicate provider: ${provider.id}`);
+        }
+        this.providers.push(provider);
+      },
+    };
+  }
+
+  start(coreStart: CoreStart): SearchServiceStart {
+    this.contextFactory = getContextFactory(coreStart);
+    return {
+      find: (term, options, request) => this.performFind(term, options, request),
+    };
+  }
+
+  private performFind(term: string, options: GlobalSearchFindOptions, request: KibanaRequest) {
+    const licenseState = this.licenseChecker!.getState();
+    if (!licenseState.valid) {
+      return throwError(
+        `GlobalSearch API is disabled because of invalid license state: ${licenseState.message}`
+      );
+    }
+
+    const context = this.contextFactory!(request);
+
+    const timeout$ = timer(this.config!.search_timeout.asMilliseconds()).pipe(map(() => undefined));
+    const aborted$ = options.aborted$ ? merge(options.aborted$, timeout$) : timeout$;
+    const providerOptions = {
+      ...options,
+      preference: options.preference ?? 'default',
+      maxResults: this.maxProviderResults,
+      aborted$,
+    };
+
+    const processResult = (result: GlobalSearchProviderResult) =>
+      processProviderResult(result, this.basePath!);
+
+    const providersResults$ = this.providers.map((provider) =>
+      provider.find(term, providerOptions, context).pipe(
+        takeInArray(this.maxProviderResults),
+        takeUntil(aborted$),
```

Comment:
Should there be a timeout on how long a provider is allowed to take before we skip it? In that case, how do we communicate to the frontend that the results may be incomplete?

Reply:
There is a timeout (see line93)

However atm the consumer has no way to know that some results were dropped. We just stop emitting them.
We could add a field in the `GlobalSearchBatchedResults` for that, but as we stop emitting when the timeout is reached, we wouldnt be emitting any new batch to specify that the completion of the observable is caused by a timeout.

Not sure what we could do here.

Reply:
Shouldn't we complete Observable in case a request was aborted? I thought we already do with `takeUntil(aborted$)`.

Reply:
We do complete the observable in case of abortion (or timeout). The (potential) issue Josh was mentioning is that there is currently no way to differentiate an observable that completed 'normally' (aka all providers completed) from one that completed because of an abortion.

Reply:
>  We could add a field in the `GlobalSearchBatchedResults` for that, but as we stop emitting when the timeout is reached, we wouldnt be emitting any new batch to specify that the completion of the observable is caused by a timeout.

Yeah we may have to emit an empty batch just to pass this information. I'm fine with punting on this, but it would be good to document this behavior / limitation somewhere (plugin README?)